### PR TITLE
[FLINK-30665][table] Planner supports row-level update

### PR DESCRIPTION
**Pull request by luoyuxia**
## What is the purpose of the change
To make planner supports row-level update.

## Brief change log
  -  Convert the `SqUpdate` to` LogicalTableModify`, and then rewrite the `LogicalTableModify` to a RelNode that insert into the sink with rows
      - If the update mode is `ALL_ROWS`,  assuming the filter is `a`, convert the update expression to `IF(a, xx, yyy)`.
      - Get the required columns by calling method SupportsRowLevelUpdate#applyRowLevelUpdate. We may rewrite the 
      LogicaTableScan to add extra meta cols to the underlying table. Then create a Project node that only select the required 
      columns.
  -  If update mode is `UPDATED_ROWS`, add a operator to set the `RowKind` to RowKind#UPDATE_AFTER in CommonExecSink
 
  - Move the method `SqlToOperationConverterTest#prepareTable` to a new class to avoid `SqlToOperationConverterTest` is too long to fail to pass check-style.

## Verifying this change
- The test in `SqlToOperationConverterTest#testUpdate` to verify update statement will be converted to `SinkModifyOperation`.
- The test in `RowLevelDeleteUpdate` to verify the plan for row-level update.
- The IT case in `UpdateTableITCase` to verify update can be performed normally.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by luoyuxia** [comment]
@lincoln-lil The pr is ready to review. Could you please help review when you're free?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 1]
```diff
@@ -1545,6 +1548,23 @@ private Operation convertDelete(SqlDelete sqlDelete) {
...
     }
 
+    private Operation convertUpdate(SqlUpdate sqlUpdate) {
+        // set it's update
+        RowLevelModificationContextUtils.setModificationType(
+                SupportsRowLevelModificationScan.RowLevelModificationType.UPDATE);
+        RelRoot updateRelational = flinkPlanner.rel(sqlUpdate);
+        // get target sink table
+        LogicalTableModify tableModify = (LogicalTableModify) updateRelational.rel;
+        List<String> targetTablePath = tableModify.getTable().getQualifiedName();
```
Comment (lincoln-lil) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:1558]: nit: use `UnresolvedIdentifier` directly to keep consistent with the delete operation

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/abilities/sink/RowLevelUpdateSpec.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A sub-class of {@link SinkAbilitySpec} that can not only serialize/deserialize the row-level
+ * update mode & updated columns to/from JSON, but also can update existing data for {@link
```
Comment (lincoln-lil) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/abilities/sink/RowLevelUpdateSpec.java:41]: 'update mode & updated columns' -> 'update mode & columns'

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/connectors/DynamicSinkUtils.java [hunk 3]
```diff
@@ -403,6 +418,59 @@ private static RelNode convertDelete(
...
+                    String.format(
+                            "Can't perform update operation of the table %s because the corresponding dynamic table sink has not yet implemented %s.",
+                            tableDebugName, SupportsRowLevelUpdate.class.getName()));
+        }
+        SupportsRowLevelUpdate supportsRowLevelUpdate = (SupportsRowLevelUpdate) sink;
+        ResolvedSchema resolvedSchema = contextResolvedTable.getResolvedSchema();
+        List<Column> updatedColumns = getUpdatedColumns(tableModify, resolvedSchema);
+        SupportsRowLevelUpdate.RowLevelUpdateInfo updateInfo =
+                supportsRowLevelUpdate.applyRowLevelUpdate(
+                        updatedColumns, RowLevelModificationContextUtils.getScanContext());
```
Comment (lincoln-lil) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/connectors/DynamicSinkUtils.java:440]: nit: add a local variable for `RowLevelModificationContextUtils.getScanContext()`

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestUpdateDeleteTableFactory.java [hunk 4]
```diff
@@ -123,6 +131,20 @@
...
 
+    private static final ConfigOption<List<String>> REQUIRED_COLUMNS_FOR_UPDATE =
+            ConfigOptions.key("required-columns-for-update")
+                    .stringType()
+                    .asList()
+                    .noDefaultValue()
+                    .withDescription("The name for the required columns in row-level update");
+
+    private static final ConfigOption<Boolean> ONLY_REQUIRE_UPDATED_COLUMNS_FOR_UPDATE =
+            ConfigOptions.key("only_require_updated_columns_for_update")
```
Comment (lincoln-lil) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestUpdateDeleteTableFactory.java:142]: 'only_require_updated_columns_for_update' -> 'only-require-updated-columns-for-update'

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestUpdateDeleteTableFactory.java [hunk 5]
```diff
@@ -123,6 +131,20 @@
...
+                    .booleanType()
+                    .defaultValue(false)
+                    .withDescription(
+                            "Whether to only require the updated columns for update statement. ");
```
Comment (lincoln-lil) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestUpdateDeleteTableFactory.java:146]: add additional comments: ', require all columns by default'

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/batch/sql/RowLevelUpdateTest.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                        String.format(
+                                "CREATE TABLE t1 (a int, b string) WITH "
+                                        + "('connector' = 'test-update-delete', 'update-mode' = '%s') ",
+                                updateMode));
+        util.verifyExplainInsert(
+                "UPDATE t SET b = 'v2' WHERE a = (SELECT count(*) FROM t1)", explainDetails);
+    }
+
+    @Test
+    public void testUpdateWithOnlyRequireUpdatedCols() {
```
Comment (lincoln-lil) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/batch/sql/RowLevelUpdateTest.java:98]: add a new case which not contains all columns in update clause, e.g., a source table with more columns here 'TABLE t (f0, f1,..., a int, b string, c double, ...)'

**Review by lincoln-lil — COMMENTED** [review]
@luoyuxia thanks for contributing this! Looks good to me overall, I've left some comments.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/utils/SqlToOperationConverterTestUtils.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+import org.apache.flink.table.catalog.GenericInMemoryCatalog;
+import org.apache.flink.table.catalog.ObjectIdentifier;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/** Utils for {@link org.apache.flink.table.planner.operations.SqlToOperationConverterTest} . */
+public class SqlToOperationConverterTestUtils {
```
[the remark this answers is not in the document]
Comment (luoyuxia) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/utils/SqlToOperationConverterTestUtils.java:36]: A good suggestion.

**Comment by luoyuxia** [comment]
@lincoln-lil Thanks for reviewing. I append two commits  to fix your comments. The first one is to address the comments for row-level update, the second one is to split the test `SqlToOperationConverterTest`.

**Review by lincoln-lil — COMMENTED** [review]
@luoyuxia thanks for your update! I have a minor comment for the test splitting: although there is no strict guideline on how to classify various sql commands, it is clearly inappropriate to classify commands such as show, load, use, etc. into ddl or dml, should we adjust it to refer to mysql's definition of ddl, dml, and put the remaining into sqlOtherOperation (probably called sqlCommand is also easy to cause ambiguity).
 Btw, since 'begin statement set' is used to assist multiple dml statements, we can also put it into dml test, WDYT?

**Comment by luoyuxia** [comment]
> @luoyuxia thanks for your update! I have a minor comment for the test splitting: although there is no strict guideline on how to classify various sql commands, it is clearly inappropriate to classify commands such as show, load, use, etc. into ddl or dml, should we adjust it to refer to mysql's definition of ddl, dml, and put the remaining into sqlOtherOperation (probably called sqlCommand is also easy to cause ambiguity). 

Agree to it. The reason I put `show`/`use` into ddl is  from the Hive's [language manual ddl  wiki](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-Show). But now, I think the hive's is not correct and suitable for all systems. So i like the idea that put them into `sqlOtherOperation`.


> Btw, since 'begin statement set' is used to assist multiple dml statements, we can also put it into dml test, WDYT?
Agree to it

**Comment by luoyuxia** [comment]
@lincoln-lil Thanks for reviewing. Now I move the test for the statements that neither belong to DDL nor DML  to `SqlOtherOperationConverterTest`

**Comment by lincoln-lil** [comment]
@flinkbot run azure

**Review by lincoln-lil — APPROVED** [review]
LGTM

**Pull request closed without merging**